[NCCL] Fix NCCL_BLOCKING_WAIT functionality with Async Error Handling

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44411 [NCCL] Fix NCCL_BLOCKING_WAIT functionality with Async Error Handling**

This basically aborts errored NCCL communicators if either blocking
wait or async error handling is enabled. Otherwise we may abort nccl
communicators where neither are enabled, and this may result in subsequent GPU
operations using corrupted data.

Differential Revision: [D23605382](https://our.internmc.facebook.com/intern/diff/D23605382/)

osalpekar committed:
[NCCL] Fix NCCL_BLOCKING_WAIT functionality with Async Error Handling

This basically aborts errored NCCL communicators if either blocking
wait or async error handling is enabled. Otherwise we may abort nccl
communicators where neither are enabled, and this may result in subsequent GPU
operations using corrupted data.

Differential Revision: [D23605382](https://our.internmc.facebook.com/intern/diff/D23605382/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Is this PR supposed to be on top of this stack: https://github.com/pytorch/pytorch/pull/43232?

osalpekar committed:
Update on "[NCCL] Fix NCCL_BLOCKING_WAIT functionality with Async Error Handling"


This basically aborts errored NCCL communicators if either blocking
wait or async error handling is enabled. Otherwise we may abort nccl
communicators where neither are enabled, and this may result in subsequent GPU
operations using corrupted data.

Differential Revision: [D23605382](https://our.internmc.facebook.com/intern/diff/D23605382/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Fix NCCL_BLOCKING_WAIT functionality with Async Error Handling"


This basically aborts errored NCCL communicators if either blocking
wait or async error handling is enabled. Otherwise we may abort nccl
communicators where neither are enabled, and this may result in subsequent GPU
operations using corrupted data.

Differential Revision: [D23605382](https://our.internmc.facebook.com/intern/diff/D23605382/)

[ghstack-poisoned]
